File: src/sync/transfer_state.rs
//! 强类型的持久传输状态、操作与错误分类。

use serde::{Deserialize, Serialize};
use thiserror::Error;

/// 持久传输任务的生命周期状态。
#[repr(i32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum TransferState {
    Pending = 0,
    Running = 1,
    WaitingForNetwork = 2,
    BackingOff = 3,
    VerifyingRemote = 4,
    RestartRequired = 5,
    Completed = 6,
    Failed = 7,
    Canceled = 8,
}

/// 持久传输任务代表的文件操作。
#[repr(i32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum TransferOperation {
    Create = 0,
    Update = 1,
    Download = 2,
    DownloadUpdate = 3,
    Delete = 4,
    Move = 5,
    Rename = 6,
    CreateFolder = 7,
}

/// 可持久的结构化传输失败类型。
#[repr(i32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum TransferErrorKind {
    Network = 0,
    Timeout = 1,
    Auth = 2,
    RateLimit = 3,
    Server = 4,
    Quota = 5,
    Permission = 6,
    Validation = 7,
    SessionExpired = 8,
    RemoteAmbiguous = 9,
    LocalChanged = 10,
    Unknown = 11,
}

/// 乐观传输状态迁移被拒绝的原因。
#[derive(Debug, Clone, PartialEq, Eq, Error)]
pub enum TransitionError {
    #[error("illegal transfer transition from {from:?} to {to:?}")]
    IllegalTransition {
        from: TransferState,
        to: TransferState,
    },
    #[error("stale transfer revision for task {task_id}: expected {expected_revision}")]
    StaleRevision {
        task_id: i64,
        expected_revision: i64,
    },
    #[error("transfer task {task_id} was not found")]
    NotFound { task_id: i64 },
    #[error("invalid stored value {value} for {field}")]
    InvalidStoredValue { field: &'static str, value: i32 },
    #[error("transfer database operation failed: {message}")]
    Database { message: String },
}

impl From<rusqlite::Error> for TransitionError {
    /// 将 SQLite 错误包装为传输迁移错误。
    fn from(error: rusqlite::Error) -> Self {
        Self::Database {
            message: error.to_string(),
        }
    }
}

/// 为持久枚举生成与整数之间的严格转换。
macro_rules! impl_persistent_enum {
    ($enum:ty, $field:literal, [$($value:literal => $variant:path),+ $(,)?]) => {
        impl From<$enum> for i32 {
            /// 将持久枚举转为数据库整数值。
            fn from(value: $enum) -> Self {
                value as i32
            }
        }

        impl TryFrom<i32> for $enum {
            /// 持久枚举整数转换失败时返回的错误。
            type Error = TransitionError;

            /// 严格将数据库整数值解析为枚举。
            fn try_from(value: i32) -> Result<Self, Self::Error> {
                match value {
                    $($value => Ok($variant),)+
                    _ => Err(TransitionError::InvalidStoredValue {
                        field: $field,
                        value,
                    }),
                }
            }
        }
    };
}

impl_persistent_enum!(
    TransferState,
    "state",
    [
        0 => TransferState::Pending,
        1 => TransferState::Running,
        2 => TransferState::WaitingForNetwork,
        3 => TransferState::BackingOff,
        4 => TransferState::VerifyingRemote,
        5 => TransferState::RestartRequired,
        6 => TransferState::Completed,
        7 => TransferState::Failed,
        8 => TransferState::Canceled,
    ]
);
impl_persistent_enum!(
    TransferOperation,
    "operation",
    [
        0 => TransferOperation::Create,
        1 => TransferOperation::Update,
        2 => TransferOperation::Download,
        3 => TransferOperation::DownloadUpdate,
        4 => TransferOperation::Delete,
        5 => TransferOperation::Move,
        6 => TransferOperation::Rename,
        7 => TransferOperation::CreateFolder,
    ]
);
impl_persistent_enum!(
    TransferErrorKind,
    "error_kind",
    [
        0 => TransferErrorKind::Network,
        1 => TransferErrorKind::Timeout,
        2 => TransferErrorKind::Auth,
        3 => TransferErrorKind::RateLimit,
        4 => TransferErrorKind::Server,
        5 => TransferErrorKind::Quota,
        6 => TransferErrorKind::Permission,
        7 => TransferErrorKind::Validation,
        8 => TransferErrorKind::SessionExpired,
        9 => TransferErrorKind::RemoteAmbiguous,
        10 => TransferErrorKind::LocalChanged,
        11 => TransferErrorKind::Unknown,
    ]
);

/// 判断 `from -> to` 是否为声明过的传输生命周期边。
pub const fn can_transition(from: TransferState, to: TransferState) -> bool {
    use TransferState::*;

    matches!(
        (from, to),
        (
            Pending,
            Running | WaitingForNetwork | RestartRequired | Failed | Canceled
        ) | (
            Running,
            WaitingForNetwork
                | BackingOff
                | VerifyingRemote
                | RestartRequired
                | Completed
                | Failed
                | Canceled
        ) | (
            WaitingForNetwork,
            Running | RestartRequired | Failed | Canceled
        ) | (BackingOff, Running | RestartRequired | Failed | Canceled)
            | (
                VerifyingRemote,
                Running
                    | WaitingForNetwork
                    | BackingOff
                    | RestartRequired
                    | Completed
                    | Failed
                    | Canceled
            )
            | (
                RestartRequired,
                Pending | VerifyingRemote | Failed | Canceled
            )
            | (Failed, Pending | RestartRequired | Canceled)
    )
}

File: shared/src/commonMain/kotlin/io/github/yuanbaobaoo/petallink/sync/TransferState.kt
package io.github.yuanbaobaoo.petallink.sync

/**
 * 传输九态状态机（对标原项目 src/sync/transfer_state.rs）
 *
 * 每次状态变更必须通过 [canTransition] 校验，并通过 CAS（state_revision）
 * 乐观锁落库（docs/04 §6 / docs/06 §10）。详见 docs/06 §九态状态机。
 *
 * 状态语义：
 * - Pending            ：已入队待执行
 * - Running            ：正在传输（上传/下载）
 * - WaitingForNetwork  ：网络不可用，等待恢复
 * - BackingOff         ：传输失败，退避等待重试
 * - VerifyingRemote    ：传输完成后核验远端（数据安全规则，docs/06 §11）
 * - RestartRequired    ：可恢复中断（需用户/引擎确认后重启）
 * - Completed          ：完成（终态）
 * - Failed             ：失败（终态，超过 MAX_AUTOMATIC_ATTEMPTS）
 * - Canceled           ：用户取消（终态）
 */
enum class TransferState {
    Pending,
    Running,
    WaitingForNetwork,
    BackingOff,
    VerifyingRemote,
    RestartRequired,
    Completed,
    Failed,
    Canceled;

    companion object {
        // ------------------------------------------------------------------
        // 合法迁移矩阵（docs/06 §can_transition）
        // key = 源状态，value = 可达目标状态集合
        // ------------------------------------------------------------------
        private val TRANSITIONS: Map<TransferState, Set<TransferState>> = mapOf(
            Pending to setOf(Running, WaitingForNetwork, RestartRequired, Failed, Canceled),
            Running to setOf(
                VerifyingRemote,        // 正常完成 → 核验远端
                WaitingForNetwork,      // 网络中断
                BackingOff,             // 可重试错误
                RestartRequired,
                Completed,
                Failed,                 // 不可重试错误
                Canceled,               // 用户取消
            ),
            WaitingForNetwork to setOf(Running, RestartRequired, Failed, Canceled),
            BackingOff to setOf(Running, RestartRequired, Failed, Canceled),
            VerifyingRemote to setOf(
                Running, WaitingForNetwork, BackingOff, RestartRequired, Completed, Failed, Canceled,
            ),
            RestartRequired to setOf(Pending, VerifyingRemote, Failed, Canceled),
            Completed to emptySet(),
            Failed to setOf(Pending, RestartRequired, Canceled),
            Canceled to emptySet(),
        )

        /**
         * 校验状态迁移是否合法。非法迁移视为 bug，必须拒绝（防状态机错乱）。
         */
        fun canTransition(from: TransferState, to: TransferState): Boolean {
            if (from == to) return true // 幂等
            return TRANSITIONS[from]?.contains(to) == true
        }

        /**
         * 终态判定（不再变化）
         */
        fun isTerminal(state: TransferState): Boolean =
            state == Completed || state == Failed || state == Canceled

        /**
         * 活跃态判定（占用传输槽位）
         */
        fun isActive(state: TransferState): Boolean =
            state == Running || state == VerifyingRemote
    }
}
